### PR TITLE
Add batch normalization folding to QAT quantizer

### PR DESCRIPTION
**Pull request by chenbohua3**
This pr adds batch normalization folding to the QAT quantizer, the core ideas are described in #3890

### CODE REVIEW

File: nni/algorithms/compression/pytorch/quantization/quantizers.py [hunk 1]
```diff
@@ -265,15 +269,15 @@ def quantize_weight(self, wrapper, **kwargs):
...
-        # if bias exists, quantize bias to uint32
-        if hasattr(wrapper.module, 'bias') and wrapper.module.bias is not None:
-            bias = wrapper.module.bias.data
-            bias_bits = 32
-            rmin, rmax = torch.min(bias), torch.max(bias)
-            module.scale, module.zero_point = update_quantization_param(bias_bits, rmin, rmax)
-            bias = self._quantize(bias_bits, module, bias)
-            bias = self._dequantize(module, bias)
-            wrapper.module.bias.data = bias
+        # # if bias exists, quantize bias to uint32
```
Comment (chenbohua3) [inline nni/algorithms/compression/pytorch/quantization/quantizers.py:272]: I will rebase this pr when #3904 is merged into master

**Comment by linbinskn** [comment]
Look good. I only have one question right now. Is there any problems If we want to export simulated model with new feature bn folding to backend execution engine such as TensorRT? For instance, during inference, conv+bn+relu will be fused into singel op by updating the conv's weight/bias parameter with bn parameters. However, currently our conv's weights have already been equal to fused weight while bn layer still exists. If the problem actual exists, maybe we can discuss an appropriate method to resolve it.

File: nni/algorithms/compression/pytorch/quantization/quantizers.py [hunk 2]
```diff
@@ -125,7 +125,7 @@ class QAT_Quantizer(Quantizer):
     http://openaccess.thecvf.com/content_cvpr_2018/papers/Jacob_Quantization_and_Training_CVPR_2018_paper.pdf
     """
 
-    def __init__(self, model, config_list, optimizer=None):
+    def __init__(self, model, config_list, optimizer=None, model_inputs=None):
```
[the remark this answers is not in the document]
Comment (chenbohua3) [reply nni/algorithms/compression/pytorch/quantization/quantizers.py]: done

**Comment by chenbohua3** [comment]
You are right. I have added some code logic to restore folded weight/bias in `export_model`.

File: nni/compression/pytorch/compressor.py [hunk 3]
```diff
@@ -525,12 +547,35 @@ def forward(self, *inputs):
...
+        return x
+
+
 class Quantizer(Compressor):
     """
     Base quantizer for pytorch quantizer
     """
 
-    def __init__(self, model, config_list, optimizer=None):
+    def __init__(self, model, config_list, optimizer=None, dummy_input=None):
```
Comment (J-shang) [inline nni/compression/pytorch/compressor.py:575]: pls also add a docstring to explain what `dummy_input` is used for.

File: nni/algorithms/compression/pytorch/quantization/quantizers.py [hunk 4]
```diff
@@ -145,8 +145,11 @@ def __init__(self, model, config_list, optimizer=None):
                     state where activation quantization ranges do not exclude a signiﬁcant fraction of values, default value is 0
                 - op_types : list of string
                     types of nn.module you want to apply quantization, eg. 'Conv2d'
+                - dummy_input : tuple of tensor
+                    inputs to the model, which are used to get the graph of the module
```
Comment (J-shang) [inline nni/algorithms/compression/pytorch/quantization/quantizers.py:149]: I think we also need to explain the graph is used to find bn after Conv. If user doesn't provide `dummy_input`, the quantizer will not fold bn.

Comment (chenbohua3) [reply nni/algorithms/compression/pytorch/quantization/quantizers.py]: done

File: nni/compression/pytorch/compressor.py (hunk 3 above)
Comment (chenbohua3) [reply nni/compression/pytorch/compressor.py:575]: I add it in `class QAT_Quantizer`.

**Review by J-shang — APPROVED** [review]

**Comment by linbinskn** [comment]
Please update content of bn folding in doc [Supported Quantization Algorithms on NNI](https://github.com/microsoft/nni/blob/master/docs/en_US/Compression/Quantizer.rst#qat-quantizer).

**Comment by chenbohua3** [comment]
the content of bn folding has been added

**Review by linbinskn — APPROVED** [review]

File: nni/algorithms/compression/pytorch/quantization/quantizers.py [hunk 5]
```diff
@@ -344,9 +367,39 @@ def export_model(self, model_path, calibration_path=None, onnx_path=None, input_
 
         return calibration_config
 
-    def fold_bn(self, config, **kwargs):
-        # TODO simulate folded weight
-        pass
+    def fold_bn(self, *inputs, wrapper):
```
Comment (QuanluZhang) [inline nni/algorithms/compression/pytorch/quantization/quantizers.py:370]: this function is `QAT_Quantizer` specific? other quantizers may have a different `fold_bn` function?

Comment (chenbohua3) [reply nni/algorithms/compression/pytorch/quantization/quantizers.py:370]: This function should also work well for other quantizers. (at least for lsq quantizer I think:) ). I will make it a common utility function in the pr that enables batch normalization folding for other quantizers.

**Review by QuanluZhang — APPROVED** [review]
